### ISSUE TITLE
#[non_exhaustive] variants are not detected to be redundant

### ISSUE THREAD

**Issue by Nadrieril** [rust-lang/rust#65157]
Say I have the following enum in crate `b`:
```rust
#![feature(non_exhaustive)]
pub enum Enum {
    Unit,
    #[non_exhaustive] Struct { x: bool }
}
```
And I match on it in crate `a`:
```rust
#![deny(unreachable_patterns)]
match Enum::Unit {
    Enum::Struct { .. } => {}
    Enum::Struct { .. } => {} // This should error but doesn't
    _ => {}
}
```
The second branch is not detected to be unreachable. This is on current nightly; [playground link (hacky)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=00568321f94cf79d6d5eece1e2bc5a14).

**Comment by Nadrieril** [comment]
The reason for this behaviour is straightforward: [this line](https://github.com/rust-lang/rust/blob/5a8fb7c24d85d0bd911ca465c4a077c7cee608ae/src/librustc_mir/hair/pattern/_match.rs#L1127) reports tuple variants with non-exhaustive fields as always useful.
This line was introduced in commit https://github.com/rust-lang/rust/commit/0d034a2e4dd9418fa41aba4ebdd1ebc31a08c9e9 . I'm not quite sure what the intended semantics are so I wouldn't know how to fix it.

**Comment by Nadrieril** [comment]
Please wait for https://github.com/rust-lang/rust/pull/65160 to be merged before fixing this; I'm dreading having to rebase on top of the fix

**Comment by arielb1** [comment]
cc @davidtwco
